### PR TITLE
update simple-icons to v5 with by-name lookup backwards compatibility

### PR DESCRIPTION
**Pull request by calebcartwright**
Implements the second option described in #6580 to update to v5 of simple-icons while maintaining backwards compatibility with how we've been handling named logos that contain spaces in the icon name/title. Note this means that icons can now be referenced via their slug or via their title, and title's containing spaces can be referenced either by escaping the spaces or replacing them with hyphens.

https://shields-staging-pr-6591.herokuapp.com/badge/foo-bar-blue?logo=linux-foundation

https://shields-staging-pr-6591.herokuapp.com/badge/foo-bar-blue?logo=linux%20foundation

https://shields-staging-pr-6591.herokuapp.com/badge/foo-bar-blue?logo=linuxfoundation

(know the icon loading isn't under the core directory but adding the label since it feels like a "core" item)

### CODE REVIEW

**Comment by chris48s** [comment]
:books: Background reading:
- https://github.com/badges/shields/issues/4273
- https://github.com/badges/shields/pull/3641

I think if we're going to (finally) enable the proper SI slugs to be used, that should be the only documented option (but the others retained for backwards compatibility). I think our original plan was to literally hard-code old --> new maps and stop generating the old style slugs but maybe we're too far gone for that and the simplest thing to do is just generate more and more of them forever. .Do you have any thoughts on the tradeoffs?

File: lib/load-simple-icons.js [hunk 1]
```diff
@@ -5,18 +5,23 @@ const { svg2base64 } = require('./svg-helpers')
...
-      default: svg2base64(
-        simpleIcons[k].svg.replace('<svg', `<svg fill="#${simpleIcons[k].hex}"`)
-      ),
-      light: svg2base64(
-        simpleIcons[k].svg.replace('<svg', `<svg fill="whitesmoke"`)
-      ),
-      dark: svg2base64(simpleIcons[k].svg.replace('<svg', `<svg fill="#333"`)),
+    const icon = originalSimpleIcons[key]
+    const title = icon.title.toLowerCase()
+    const legacyTitle = title.replace(/ /g, '-')
```
Comment (chris48s) [inline lib/load-simple-icons.js:16]: should this be `title.toLowerCase().replace(/ /g, '-')` ?

**Comment by calebcartwright** [comment]
> Background reading:

Thanks for this. Vaguely remember reading those back in the day but the dots hadn't fully connected for me until now.

> I think if we're going to (finally) enable the proper SI slugs to be used, that should be the only documented option (but the others retained for backwards compatibility)

SI have forced our hand on this with the v5 update. We'd have to really bend over backwards to _not_ support the proper slugs, and I can't imagine why we'd do extra work to avoid that. If we intend to continue supporting icon-by-name as well (regardless of fully or partially via hard coded table), then I guess I'd be fine not documenting/publicizing it.

> I think our original plan was to literally hard-code old --> new maps and stop generating the old style slugs but maybe we're too far gone for that and the simplest thing to do is just generate more and more of them forever. . Do you have any thoughts on the tradeoffs?

At this point I'd be opposed to attempting to maintain a hard coded map of old slugs to the current set.

There'd be far too many, and then if/when the slugs are updated in SI we'd have to go update our maps. We'd also have to deal with questions/requests to maintain others (technically if we were to maintain such a mapping we _could_ include entries for all these [updated](https://github.com/badges/shields/issues/6590) to prevent logos from disappearing on existing badges when we deploy a major version update. I know we aren't obligated to do so, just noting I'm not keen on having to explain-to/debate-with users why we're mapping some but not all.

I'd also worry about the impact on the user experience. At this point we certainly have users that have become accustomed to current by-name reference using the `-` or `%20` delimiters, and I think it'd be really confusing as a user if I've always used that pattern, e.g. `?logo=visual-studio-code`, and that pattern works for a bunch of other logos (assuming hard coded mapping), but when I try to apply that same pattern to some new `Foo Bar` logo it doesn't work, since in this hypothetical we wouldn't have mappings for new logos in the old table, and wouldn't directly support the name reference. Just seems it'd be really confusing/magical why the pattern would work for some/most but not all logos.

At the end of the day, I think it's inevitable that we'll start supporting the by-slug references directly as part of the v5 update, so IMO the question is really whether we want to also continue supporting by-name (and if not then if/how on the BC front).

Also, I'm not quite sure what the concerns or issues are with also supporting the by-name pattern given the v5 context. Feels like the reported issues have historically been driven by challenges associating to slug (no longer an issue IMO with the change of the exports in v5), users attempting and failing to use the actual slug in cases where our coercion failed to match the actual slug, couple quirky names (apostrophes/backticks), and our own locally maintained logos. Anything you can share/remind me of relative to why supporting both would be problematic?

File: lib/load-simple-icons.js (hunk 1 above)
Comment (calebcartwright) [reply lib/load-simple-icons.js:16]: let me know if i've misunderstood, but `title` would already be lower cased as part of the initial assignment on the preceding line wouldn't it?

**Comment by developStorm** [comment]
I was working on [a project](https://github.com/developStorm/simple-badges) that shows all the Simple Icons with corrensponding Shields IO badge and experenced a hard time to find the correlation between slug/title of Simple Icons with logo option of Shields IO until I saw this PR (looks like I just happened to stuck in a major version update). 

I've successfully integrated the test build of Shields above into my project, and it seems to be working well. Hope this could be deployed soon!

File: lib/load-simple-icons.js (hunk 1 above)
Comment (developStorm) [reply lib/load-simple-icons.js:16]: There may be a compatibility problem here - the `title` may have changed in 5.0 as well. I tried before to use 5.0 icons and do the title transformation as what has been done here, but I actually had to do some extra processing to get it to work properly.

An example would be `Node.JS` (actually almost all badges with dot in title have this problem):

https://shields-staging-pr-6591.herokuapp.com/badge/foo-bar-blue?logo=node-dot-js


https://shields.io/badge/foo-bar-blue?logo=node-dot-js



https://shields-staging-pr-6591.herokuapp.com/badge/foo-bar-blue?logo=node.js


https://shields.io/badge/foo-bar-blue?logo=node.js

Comment (calebcartwright) [reply lib/load-simple-icons.js:16]: @developStorm - you're referencing breaking changes that SimpleIcons made that's separate from the name vs. slug reference (more details in #6590). We don't attempt to maintain backwards compatibility when SI change their own slugs, and that's unrelated to this thread so going to hide these last two comments as off topic to keep the thread focused

**Comment by chris48s** [comment]
> At this point I'd be opposed to attempting to maintain a hard coded map of old slugs to the current set.

All good points :+1: Being ~2 years further down the path we are already on is significant.

> Anything you can share/remind me of relative to why supporting both would be problematic?

I think the only downside with generating two sets of our own additional identifiers forever as well as the SI-provided slugs would be if there is any way we might generate a conflict or collision between an ID we've generated and one of the proper SI slugs. It looks super-unlikely that would happen but its hard to commit to saying it will never happen with any future release of SI.

For belt-and-braces we could throw an exception if we generate `title` or `legacyTitle` and it is in `Object.keys(originalSimpleIcons).filter(slug => slug !== key)`. If it never happens, great. If it does, the tests fail when we upgrade to the release that causes it and we defer working out what to do about it to us-from-the-future.

**Comment by calebcartwright** [comment]
> It looks super-unlikely that would happen but its hard to commit to saying it will never happen with any future release of SI.

Interesting scenario. The only extra slugs we'd be generating with this proposal is the lowercased version of the title, and the lowercased version of the title with the spaces replaced with hyphens. As of v5 SI no longer permit hyphens in the slugs (and never have had whitespace AFAIK), so the only path I could see for this to occur would be something a case where the title of one icon randomly matches the slug of another in the exports, e.g. something like this

```js
{
  dog: { title: 'Cat', slug: 'dog', ... },
  cat: { title: 'Duck', slug: 'cat', ... },
  ...
}
```

Guess it's not 100% impossible even if it does feel like a stretch, but yes we could certainly include a check for this scenario

**Comment by ericcornelissen** [comment]
Hi all :wave: One of the core maintainers of Simple Icons here. I'm sorry for all the trouble we're causing :sweat_smile:

I wanted to chime in to clarify some things regarding the `slug` value. The `slug` value is specifically intended to avoid problems due to brands with the same name or auto-generated slug (for example [we have two icons both for a brand called "Hive"](https://github.com/simple-icons/simple-icons/blob/7cdc8ac6a3ff0f46857ed490661298b6cfe7d90d/_data/simple-icons.json#L3887-L3897)) So, the `slug` is either [derived from the title deterministically](https://github.com/simple-icons/simple-icons/blob/cae25487b37975758a5753e8711681a179110586/scripts/utils.js#L17-L32), indeed without whitespace and now also without hyphens, or it is custom, unique, and hardcoded in our repository in which case it will always contain an `_`.

So, in short, the above scenario would never occur. Supporting our slug value is a must if you want your users to be able to use all our icons; but using the title or hyphenated title should also always work except for the rare case where multiple brands share a name.

**Comment by calebcartwright** [comment]
No worries @ericcornelissen and thanks for the additional context! Long term I think the changes in v5 will actually be better for us and address some long standing problems. Just a matter of whether/how much we want to do to prevent logos disappearing on badges for some of our existing users :smile:

**Comment by DenverCoder1** [comment]
:+1: Version can be bumped to `5.1.0` as well now

**Comment by calebcartwright** [comment]
@chris48s

Have kept the by-name support in there, but updated docs and the site to remove those mentions and point everyone to the slugs. Let me know if you'd still like add some handling for existing functions. The case @ericcornelissen shared with Hive is a great one, as it demonstrates why we wouldn't want throw an exception on encountering a title that maps to an existing key. At the moment, such cases will mean that we process that particular icon twice on loading, but the end result would still be the same correct icon mapping.

That's obviously doing some unnecessary work and I'm tempted to add the key for presence in _our_ map to avoid it, though held off for the moment since this is a one time/loadup not on a hotpath and because I feel like such an addition would need some additional inline commentary to provide context, which would have us trending towards having more lines of comments than javascript :laughing: 

(also went ahead and updated to v5.1)

File: lib/load-simple-icons.js [hunk 2]
```diff
@@ -5,18 +5,27 @@ const { svg2base64 } = require('./svg-helpers')
...
+    icon.base64 = {
+      default: svg2base64(icon.svg.replace('<svg', `<svg fill="#${icon.hex}"`)),
+      light: svg2base64(icon.svg.replace('<svg', `<svg fill="whitesmoke"`)),
+      dark: svg2base64(icon.svg.replace('<svg', `<svg fill="#333"`)),
     }
+    simpleIcons[key] = icon
+    simpleIcons[title] = icon
+    simpleIcons[legacyTitle] = icon
```
Comment (chris48s) [inline lib/load-simple-icons.js:28]: Shall we change the order here.
i.e: if we're going to say the SI slug is the canonical ID and the others are fallbacks, then in the case of a collision like "hive" then if we set the SI slug last then that value is guaranteed to "win" the clash here.

Comment (calebcartwright) [reply lib/load-simple-icons.js]: Good shout. Actually needed to incorporate the key check after all as originally proposed (just skipping instead of throwing exception). The order of the assignment within a given iteration doesn't matter so much, but more about needing the conditional checks otherwise the value would be determined by the last iteration that contains a colliding key. 32554a2a5b7471630d2dff5f61dd9e443b437fef

**Review by chris48s — APPROVED** [review]
:+1:  Lets ship it :ship:

**Comment by 0f3eacd3** [comment]
Hello, is there an estimated time for deploying this on production? 🙏
